### PR TITLE
PS Compiler 0.14 compatability

### PR DESCRIPTION
**Pull request by newlandsvalley**
* Added spago although retained pulp in ```package.json``` to allow the travis build.
* Replaced ```StrongCheck``` with ```QuickCheck```

The ```Semiring``` check tests now crash so I've commented them out for the time being.  The other tests all pass. 
.

### CODE REVIEW

**Comment by anttih** [comment]
I suggest we add a dev dependency on bigints and use that in the tests, it won’t overflow. We can later decide to tackle #33 and change bigints to a runtime dep.

**Comment by newlandsvalley** [comment]
Do you mean, then, that the tests will test ```Ratio``` but not ```Rational```?

**Comment by anttih** [comment]
Yes, the tests would use `Ratio BigInt`. I think we can then also remove that small int stuff from the tests.

**Comment by newlandsvalley** [comment]
OK.  Shall I take a look at it?

**Comment by anttih** [comment]
Yeah, go ahead!

**Comment by newlandsvalley** [comment]
The travis build fails only because there is no 0.14 compatible version of purescript-bigints yet published.  However you can both build and test with spago.

File: package.json [hunk 1]
```diff
@@ -3,10 +3,15 @@
...
+    "spago-test": "spago -x ./test.dhall test"
   },
   "devDependencies": {
-    "pulp": "^12.3.0",
-    "purescript": "^0.12.0"
+    "pulp": "^15.0.0",
+    "purescript": "^0.14.0"
+  },
+  "dependencies": {
+    "big-integer": "^1.6.48"
```
Comment (anttih) [inline package.json:13]: This should probably be a `devDependency` as well.

File: spago.dhall [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{-
+Welcome to a Spago project!
+You can edit this file as you like.
+-}
+{ name = "rationals"
+, dependencies = [ "console", "effect", "integers", "psci-support" ]
```
Comment (anttih) [inline spago.dhall]: We don't need `console` and `effect` I don't think.

Comment (anttih) [reply spago.dhall]: Not sure about `psci-support` either. Or do we need it for the repl to work?

File: test/Main.purs [hunk 3]
```diff
@@ -5,18 +5,21 @@ import Prelude
...
+import Test.QuickCheck.Arbitrary (class Arbitrary)
+import Test.QuickCheck.Gen (Gen, chooseInt, suchThat)
+import Test.QuickCheck.Laws (checkLaws)
+import Test.QuickCheck.Laws.Data as Data
+import Test.QuickCheck.Laws.Data.Field (checkField) as DataField
 
 import Type.Proxy (Proxy(Proxy))
 
-newtype TestRational = TestRational Rational
+type BiggerRational = Ratio BigInt
```
Comment (anttih) [inline test/Main.purs]: I'd prefer to remove this type alias, I think it's easier to read without it.

File: test/Main.purs [hunk 4]
```diff
@@ -26,29 +29,24 @@ derive newtype instance ringTestRational :: Ring TestRational
...
 
-int :: Gen Int
-int = chooseInt (-999) 999
-
-nonZeroInt :: Gen Int
-nonZeroInt = int `suchThat` notEq 0
-
-newtype SmallInt = SmallInt Int
+bigint :: Gen BigInt
+bigint = fromInt <$> chooseInt (-999) 999
```
Comment (anttih) [inline test/Main.purs]: You can now remove the `chooseInt` and just use `arbitrary` to get an arbitrary `Int`. I think the `SmallInt` was there because of the overflow.

**Review by anttih — COMMENTED** [review]
Looks good, just a few comments.

File: test/Main.purs (hunk 3 above)
Comment (newlandsvalley) [reply test/Main.purs]: agreed

File: test/Main.purs (hunk 4 above)
Comment (newlandsvalley) [reply test/Main.purs]: good point!

File: package.json [hunk 5]
```diff
@@ -3,10 +3,16 @@
   "scripts": {
     "postinstall": "bower install",
```
Comment (anttih) [inline package.json:4]: Maybe just use `spago` by default so this could just be `spago install`.

File: package.json [hunk 6]
```diff
@@ -3,10 +3,16 @@
...
     "build": "pulp build",
-    "test": "pulp test"
```
Comment (anttih) [inline package.json:6]: Let's remove these `pulp` versions and just use `spago` for both `build` and `test`.

File: package.json [hunk 7]
```diff
@@ -3,10 +3,16 @@
...
   },
   "devDependencies": {
-    "pulp": "^12.3.0",
-    "purescript": "^0.12.0"
+    "pulp": "^15.0.0",
+    "purescript": "^0.14.0",
+    "big-integer": "^1.6.48"
+  },
+  "dependencies": {
+    "big-integer": "^1.6.48"
```
Comment (anttih) [inline package.json]: This shouldn't be a proper dependency, it's enough that it's in the `devDependencies`.

File: package.json [hunk 8]
```diff
@@ -3,10 +3,16 @@
...
     "build": "pulp build",
-    "test": "pulp test"
+    "test": "pulp test",
+    "spago-build": "spago build",
+    "spago-test": "spago -x ./test.dhall test"
   },
   "devDependencies": {
-    "pulp": "^12.3.0",
-    "purescript": "^0.12.0"
+    "pulp": "^15.0.0",
```
Comment (anttih) [inline package.json:11]: Let's remove `pulp` from here and add replace it with `spago`.

**Review by anttih — COMMENTED** [review]
I still added some comments about the deps and the build commands.

File: package.json (hunk 8 above)
Comment (newlandsvalley) [reply package.json:11]: The thing is, if you do this, you can't publish

Comment (newlandsvalley) [reply package.json:11]: Also,  of course, travis would fail.  Would you prefer to move to github actions?

Comment (anttih) [reply package.json:11]: Ah ok. Maybe we keep `pulp` then in the deps. No need to move to github actions in this PR.

Comment (newlandsvalley) [reply package.json:11]: I agree.  I'll just fix the big-integer dependency.  It means we can't complete this work until there's a new PS 0.14 release of purescript-bigints.

File: package.json (hunk 6 above)
Comment (anttih) [reply package.json:6]: Could you still make the `spago` ones the default ones so we can pass the build on travis?

Comment (newlandsvalley) [reply package.json:6]: Remember that as it stands, ```travis.yml``` will not build it if we do this because it still uses pulp.  I'm very happy to make this change but I'd prefer to leave editing ```travis.yml``` to you - it's such a black art. Do you still want me to go ahead?

**Comment by anttih** [comment]
Thank you for working with this! I'll fix the CI stuff.

**Comment by newlandsvalley** [comment]
Many thanks.  It's been a pleasure!
